### DISCUSSION TITLE
`EIP-1193` typing for `window.ethereum`

### DISCUSSION THREAD

**Discussion by KipCrossing** [wevm/viem#977] [Question] [answered]
I'm trying to add typing for `window.ethereum`. I found setting the equivalent for `ethers`, in the `global.d.ts` (svelte) file that looks like:

```ts
/// <reference types="svelte" />

import { ethers } from "ethers";
import { BrowserProvider, Eip1193Provider } from "ethers/types/providers";

declare global {
  interface Window {
    ethereum: Eip1193Provider & BrowserProvider;
  }
}
```

I've tried using the `Eip1193Provider` type from `viem`:

https://github.com/wagmi-dev/viem/blob/0f7a0eb39c129ea7abc31ad328e38b75358f44ad/src/types/eip1193.ts#L24

But It's lacking return types; eg on `.request({ method: "eth_getBalance", params: [account, "latest"] })` returns unknown.

Are there additional types that can be added from `viem`?

*Context; for me, it's quality of life thing as I don't really know the `EIP-1193` spec*

**Comment by tmm** [comment, marked as answer]
Types work fine for me in [this TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAcgG7AKYgPQHdgDsAmEWCAUCQMYS4DO8Uq1ArgDbwC8cAhlp8PDgSIA6VDAAWqeoxAB+IfQCOjBjAAUAbxJw4IUWIj4AXIj0B9AOaiAQp2adc5VAgA0WuGE5ROIasYDaCAAMAB4uiHYwKggAuq4AvgCUJPRMrCQYGHAAejIkQA).

```ts
import 'viem/window'

const result = await window.ethereum?.request({
  method: 'eth_getBalance',
  params: ['0x', 'latest'],
})
result
// ^? const result: `0x${string}` | undefined
```

**Reply by KipCrossing** [reply]
Ok I didn't know you could simply:

```ts
import 'viem/window'
```

It has different(better) behavior then doing it via the `global.d.ts` file. 

I think it would be super handy for devs implementing `viem` and making dApps. Want me to add this into the docs?

**Reply by tmm** [reply]
Go for it! Maybe add it to the [TypeScript](https://viem.sh/docs/typescript.html) section? Also, open to other location.
